### ISSUE TITLE
Some non-terminal paths erroneously being added to terminalPaths by findAreaPaths()

### ISSUE THREAD

**Issue by JohnMTorgerson** [JohnMTorgerson/Viking#1]
See the following screenshot:

![screen shot 2015-09-07 at 3 02 43 am](https://cloud.githubusercontent.com/assets/6146526/9712064/931604b8-550d-11e5-9f9e-3d5aa21efe48.png)

Note the two visible paths presented in the chat window on the screenshot. The first one ends on Caqueta, which is not a terminal country, as it has a valid neighbor. The path explored right after it follows that neighbor. No path should end on Caqueta, since it does have a valid neighbor.

EDIT: The continent in question is Colombia, in the lower right hand corner of the map.

**Comment by Atelaes** [comment]
I remain at a loss to explain this behaviour.  Could you check and see if the second country on the shown path belongs to Viking or not?  I thought it did at first, but am now uncertain.  Also, you had mentioned that some of resultant arrays had empty elements in them.  Could you give an example or elaborate on this phenomenon a bit?  Still haven't managed to get Lux working on my laptop (sent another email to support), so I can't compile and test myself.

**Comment by Atelaes** [comment]
We found that the branches are being determined correctly, but then when multiple branches are concatenated together, one is sometimes changed.  It seems as though when the first terminal branch is an immediate offspring of the common ancestor, its final element is changed to the element of the same index in the last terminal branch.  This determination is not absolute.  As of yet, we have no idea how or why this is happening.

**Comment by JohnMTorgerson** [comment]
Screenshots to demonstrate the above:

![screen-shot-2015-09-10-at-2 52 26-am](https://cloud.githubusercontent.com/assets/6146526/9783376/720cda9a-5769-11e5-93e1-87c685084ae0.png)
![screen-shot-2015-09-10-at-2 52 45-am](https://cloud.githubusercontent.com/assets/6146526/9783375/720ac75a-5769-11e5-8a5f-853a0dd29b7d.png)

**Comment by JohnMTorgerson** [comment]
I added a random number at the beginning of each terminal path (overwriting the first element) in order to try to track what is being changed when the bug occurs. We had assumed it was just the last element, but I wondered if it might not be the whole array being replaced; since all the elements except the last one that's being erroneously changed are already identical to the sister/niece path that's overwriting it, we can't tell which. Once I added the random number, I confirmed that it _is_ replacing the whole array, as you can see in this screenshot:

![screen-shot-2015-09-10-at-9 01 12-pm](https://cloud.githubusercontent.com/assets/6146526/9805714/02bdf1d0-5802-11e5-8d22-2be98bd7a3dc.png)

"Great, we narrowed it down!", you say? Not so fast. Look at this next example, which occurred in the same data set:

![screen-shot-2015-09-10-at-9 00 20-pm](https://cloud.githubusercontent.com/assets/6146526/9805739/4aad3046-5802-11e5-80fd-674cce3f62ad.png)

In this case, as you can see, the random number at the beginning of the path was not overwritten, while the last element in the array _was_. So, unless I've misread what's happening (which I sincerely hope is the case), it seems that only in some cases is the whole array being overwritten, while in other cases it's only the last element. This is not what I was expecting to find. I'm completely at a loss to explain it.

**Comment by JohnMTorgerson** [comment]
So I _think_ I may have fixed it. Here's a screenshot. I'm pretty sure we always saw the bug in this scenario, right?:

![screen-shot-2015-09-19-at-9 45 44-pm](https://cloud.githubusercontent.com/assets/6146526/9979021/d5b062c6-5f19-11e5-86ec-bcf3d89ee672.png)

I realized that the only place we weren't copying the history array into a new array (newHistory) was when we added it directly to terminalPaths when we found no valid neighbors. I'm pretty sure that Java was putting a reference to the array (instead of a copy of it) into terminalPaths which was then getting changed by the mother function (where it's called newHistory, remember) on subsequent iterations of the for loop. So all I did was make a copy of the history array and add that to terminalPaths instead. We should run some more tests to be certain, but I haven't been able to replicate the bug since then.
